src/ql/pass/opt/const_prop/detail/platform_functions.h

```diff
@@ -0,0 +1,54 @@
+#pragma once
+
+/*
+ * Define platform functions using the venerable X macro mechanism
+ *
+ * We only need to implement libqasm functions which are overridden by OpenQL, and - since libqasm supports
+ * polymorphism - only for the parameter types overridden (i.e. bit and int).
+ * Also see the comment in arch::cc::pass::gen::vq1asm::detail::codegen.h::handle_set_instruction() on how platform
+ * functions come into existence.
+ *
+ * Return types (R_*):
+ * - R_B: bit
+ * - R_I: integer
+ * Parameter types (P_*):
+ * - P_B: bit literal
+ * - P_I: integer literal
+ *
+ * Based on:
+ * - deps/libqasm/src/cqasm/func-gen/funcgen.cpp
+ * - <build>/deps/libqasm/src/cqasm/cqasm-v1-functions-gen.cpp
+ * - src/ql/arch/cc/pass/gen/vq1asm/detail/functions.cc
+ *
+ * Note: to have better guarantees on consistency, we might also use this file in ir/old_to_new.cc to create the
+ * platform functions, and extend it to be usable for ir/operator_info.cc
```

Comment:
Somehow this set of functions is not a proper part of the IR now.
More and more passes will need this list so standardization is required.
And it needs to be at least a superset of what all OpenQL targets support.
And then there needs to be pass for each target that checks and enforces the entry conditions of the code generator (which can have run-time equivalents, etc.).
There maybe even functions to which calls are created by the compiler and that are not created because calls were found in a cqasm file; this certainly is true when you consider that there are multiple cqasm versions, but also when another language (e.g. OpenQASM) would be read by OpenQL.
So the function list needs to be standardized as part of the IR of OpenQL; and we need to be able to distinguish such standard functions from regular functions.
I guess the above is not comment on this code but deserves a proper issue.

Reply:
duplicate